server/src/com/cloud/storage/VolumeApiServiceImpl.java

```diff
@@ -2639,22 +2639,23 @@ private int getMaxDataVolumesSupported(UserVmVO vm) {
         return maxDataVolumesSupported.intValue();
     }
 
-    private Long getDeviceId(long vmId, Long deviceId) {
+    private Long getDeviceId(UserVmVO vm, Long deviceId) {
         // allocate deviceId
-        List<VolumeVO> vols = _volsDao.findByInstance(vmId);
+        int maxDataVolumesSupported = getMaxDataVolumesSupported(vm);
```

Comment:
I think this is a good question. I would add that the ID `0` is reserved for the root device.

So, I add the question if the `maxDataVolumesSupported` already accounts for the one already reserved for the root disk. 

For instance, when configuring `getMaxDataVolumesSupported(vm)` with `6` for the hypervisor of the VM. Does that mean that the VM can have up to 7 devices? 1 root (id `0`), 1 CD-ROM (id `3`) and other 5 for extra disks/volumes.